alpa/serialization.py

```diff
@@ -0,0 +1,166 @@
+        return msgpack.unpackb(data)
+    elif code == _MsgpackExtType.replicated_distarray:
+        return msgpack.unpackb(data)
+    return msgpack.ExtType(code, data)
+
+
+def save_checkpoint(ckpt_dir: Union[str, os.PathLike], target: PyTree,
+                    step: int):
+    """Save a checkpoint of the `target` to `path`. 
+
+        Similar to flax.training.checkpoints.save_checkpoint, but support DistributedArrays 
+        and ReplicatedDistributedArray in alpa.
+        # TODO: copy all the safe-saving stuff from 
+        https://flax.readthedocs.io/en/latest/_modules/flax/training/checkpoints.html#save_checkpoint
+
+        Args:
+           ckpt_dir: str or pathlib-like path to store checkpoint directories in.
+           target: serializable flax object, usually a trainState
+           step: training step number or other metric number
+    """
+    state_dict = to_state_dict(target)
+    os.makedirs(ckpt_dir, exist_ok=True)
+    ckpt_path = os.path.join(ckpt_dir, f"checkpoint_{step}")
+    with gfile.GFile(ckpt_path, 'wb') as fp:
+        fp.write(
+            msgpack.packb(state_dict,
+                          default=_msgpack_ext_pack_wrapper(ckpt_dir),
+                          strict_types=True))
+
+
+class LoadInfo:
+    """
+    A wrapper for the loading information.
+    """
+    def __init__(self, 
+                 avals: Sequence[ShapedArray], 
+                 meshes: Sequence[PhysicalDeviceMesh], 
+                 specs: Sequence[ShardingSpec]):
+        assert len(avals) == len(meshes)
+        assert len(meshes) == len(specs)
+        self.avals = avals
+        self.meshes = meshes
+        self.specs = specs
+
+    def add_replica(self, aval, mesh, spec):
+        self.avals.append(aval)
+        self.meshes.append(mesh)
+        self.specs.append(spec)
+
+    def get_info(self):
+        if self.is_replicated():
+            return zip(self.avals, self.meshes, self.specs)
+        else:
+            return self.avals[0], self.meshes[0], self.specs[0]
+
+    def is_replicated(self):
+        return len(self.avals) > 1
+
+
+def restore_checkpoint(ckpt_dir: Union[str, os.PathLike], step: int, target: PyTree, load_info: PyTree):
+    """Restore the specified checkpoint from `path`. 
+
+        Similar to flax.training.checkpoints.load_checkpoint, 
+        but support DistributedArrays and ReplicatedDistributedArray in alpa.
+        # TODO: copy all the safe-loading stuff from 
+        https://flax.readthedocs.io/en/latest/_modules/flax/training/checkpoints.html#restore_checkpoint
+
+        Args:
+            ckpt_dir: directory of checkpoints to restore from.
+            step: step number to load.
+            target: matching object to rebuild via deserialized state-dict.
+            load_info: shardingSpec and deviceMesh allocation info for loading.
+    """
+    ckpt_path = os.path.join(ckpt_dir, f"checkpoint_{step}")
+    with gfile.GFile(ckpt_path, 'rb') as fp:
```

Comment:
Actually, if the only dependency to gfile is here, can we try to replace this gfile with python's native file and see whether the code still work for our cases?

Reply:
Sure, I will try.